### PR TITLE
Games: Add a 2048 game

### PR DESCRIPTION
**Pull request by alimpfard**
> Game or something, idk I'm not a game dev

shot-2020-08-09_02-56-17

Featuring: 
- Badly made icons
- One of the most boring games ever
- Code that is too functional

Starring:
- A Shell bug (#3051)

### CODE REVIEW

**Comment by alimpfard** [comment]
_(sidenode: I did this just to mess around with LibGUI, it's mostly here to support #3051 than as a "please merge this")_

File: Games/2048/main.cpp [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+#include <LibGUI/BoxLayout.h>
+#include <LibGUI/Button.h>
+#include <LibGUI/Menu.h>
+#include <LibGUI/MenuBar.h>
+#include <LibGUI/Window.h>
+#include <stdio.h>
+
+int main(int argc, char** argv)
+{
+    if (pledge("stdio rpath wpath cpath shared_buffer accept cpath unix fattr", nullptr) < 0) {
```
Comment (bgianfo) [inline Games/2048/main.cpp:40]: Do you really need wpath/cpath/accept?

File: Games/2048/main.cpp [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    app_menu.add_action(GUI::Action::create("New game", { Mod_None, Key_F2 }, [&](auto&) {
+        game.reset();
+    }));
+    app_menu.add_action(GUI::CommonActions::make_quit_action([](auto&) {
+        GUI::Application::the()->quit();
+    }));
+
+    auto& help_menu = menubar->add_menu("Help");
+    help_menu.add_action(GUI::Action::create("About", [&](auto&) {
+        GUI::AboutDialog::show("2048", Gfx::Bitmap::load_from_file("/res/icons/32x32/app-2048.png"), window);
```
Comment (bgianfo) [inline Games/2048/main.cpp:74]: unviel /res/ and then unveil(nullptr, nullptr) ?

File: Games/2048/main.cpp (hunk 1 above)
Comment (alimpfard) [reply Games/2048/main.cpp:40]: Hmmm...for that matter, why does Snake need them?

Comment (awesomekling) [reply Games/2048/main.cpp:40]: They are used by the inspector feature's RPC socket.
I think you can drop `cpath unix fattr` after creating an event loop and connecting to the WindowServer.

**Comment by awesomekling** [comment]
Fourty -> Forty :)

**Comment by alimpfard** [comment]
> Fourty

lol oops

**Comment by alimpfard** [comment]
Figured I might as well make it look less bad across different themes:
shot-2020-08-09_17-54-53
shot-2020-08-09_17-54-38

**Comment by awesomekling** [comment]
Comfy :)

File: Games/2048/2048.cpp [hunk 3]
```diff
@@ -0,0 +1,351 @@
...
+    for (size_t i = 0; i < board.size(); ++i) {
+        auto& row = new_board[i];
+        row.clear_with_capacity();
+        row.ensure_capacity(result_column_count);
+        for (auto& entry : board) {
+            row.append(entry[i]);
+        }
+    }
+
+    return move(new_board);
```
Comment (awesomekling) [inline Games/2048/2048.cpp:113]: I think the compiler will automatically move return values and you don't need to explicitly `move()` them.

File: Games/2048/2048.h [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#include <AK/NonnullRefPtrVector.h>
+#include <LibGUI/Widget.h>
+
+class TwentyFortyEightGame : public GUI::Widget {
```
Comment (awesomekling) [inline Games/2048/2048.h:32]: `final`

Comment (alimpfard) [reply Games/2048/2048.h:32]: But what if someone wanted to make a 4096 game? inheriting this is _clearly_ the right path forward :P

**Comment by awesomekling** [comment]
Very chill game. The scoring system is kinda lame, I feel like we could be giving out points for every successful tile merge, not just for the highest tile number reached :)

**Comment by alimpfard** [comment]
that's how I remember the score being, tbh

> successful tile merge

Sure, I guess?

**Pull request merged**